larq/optimizers_v2.py

```diff
@@ -5,88 +5,75 @@
+        takes one `tf.Variable` as argument and returns `True` if the optimizer should
+        train that variable, e.g. `pred(var) === True`.
+    default: a `tf.keras.optimizers.Optimizer` to be applied to any variable not claimed
+        by any other optimizer.
     """
 
-    def __init__(self, masked_optimizers, name="Group"):
+    def __init__(self, pred_opt_pairs, default, name="optim_case"):
         super().__init__(name=name)
 
-        for i, masked_opt in enumerate(masked_optimizers):
-            if not isinstance(masked_opt, MaskedOptimizer):
+        for i, (pred, opt) in enumerate(pred_opt_pairs):
+            if not isinstance(opt, tf.keras.optimizers.Optimizer):
                 raise TypeError(
-                    f"Expected `MaskedOptimizer` at index {i} but got `{type(masked_opt)}`."
+                    f"Expected `tf.keras.optimizers.Optimizer` at `pred_opt_pairs[{i}][1]` but got `{type(opt)}`."
                 )
+        self.pred_opt_pairs = pred_opt_pairs
 
-        self.masked_optimizers = masked_optimizers
+        if (
+            not isinstance(default, tf.keras.optimizers.Optimizer)
+            and default is not None
+        ):
+            raise TypeError(
+                f"Expected `tf.keras.optimizers.Optimizer` for `default` but got `{type(fallback_optimizer)}`."
+            )
+        self.default = default
 
     def __getattr__(self, name):
-        if name == "lr":  # TODO: Which optimizer's LR should handle this?
+        if name == "lr":  # TODO: Return list of learning rates
             raise NotImplementedError()
         return super().__getattr__(name)
 
     def apply_gradients(self, grads_and_vars, name=None):
-        opt_grads_and_vars = [[] for _ in range(len(self.masked_optimizers))]
+        opt_grads_and_vars = [[] for _ in range(len(self.pred_opt_pairs))]
+        default_grads_and_vars = []
 
         for grad, var in grads_and_vars:
-            for i, masked_opt in enumerate(self.masked_optimizers):
-                # TODO: Check partition condition?
-                if masked_opt.mask_fn is None or masked_opt.mask_fn(var):
+            num_opts = 0
+            for i, (pred, opt) in enumerate(self.pred_opt_pairs):
+                if pred(var):
                     opt_grads_and_vars[i].append((grad, var))
+                    num_opts += 1
 
-        train_ops = []
-        for masked_opt, grads_and_vars in zip(
-            self.masked_optimizers, opt_grads_and_vars
-        ):
-            train_ops.append(
-                masked_opt.optimizer.apply_gradients(grads_and_vars, name=name)
+            if num_opts == 0:
+                default_grads_and_vars.append((grad, var))
+            if num_opts > 1:
+                raise ValueError(f"Variable `{var}` claimed by multiple optimizers.")
+
+        if len(default_grads_and_vars) > 0 and self.default is None:
+            raise ValueError(
```

Comment:
Would raise a warning, as this may be intended behaviour

Reply:
Changed to warning in 678590c.